File: 111. Minimum Depth of Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
  def minDepth(self, root):
    """
    :type root: TreeNode
    :rtype: int
    """
    def dfs(current, depth):
      if current == None:
        return depth - 1
      elif current.left == None and current.right == None:
        return depth
      elif current.left != None and current.right == None:
        return dfs(current.left, depth + 1)
      elif current.left == None and current.right != None:
        return dfs(current.right, depth + 1)

      leftDepth = dfs(current.left, depth + 1)
      rightDepth = dfs(current.right, depth + 1)
      if leftDepth < rightDepth:
        return leftDepth
      return rightDepth
    
    return dfs(root, 1)

File: 111. Minimum Depth of Binary Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  int dfs(TreeNode current, int depth){
    if (current == null){
      return depth - 1;
    }
    else if (current.left == null && current.right == null){
      return depth;
    }
    else if (current.left != null && current.right == null){
      return dfs(current.left, depth + 1);
    }
    else if (current.left == null && current.right != null){
      return dfs(current.right, depth + 1);
    }

    int leftDepth = dfs(current.left, depth + 1);
    int rightDepth = dfs(current.right, depth + 1);
    if (leftDepth < rightDepth){
      return leftDepth;
    }
    return rightDepth;
  }
  public int minDepth(TreeNode root) {
    return dfs(root, 1);
  }
}